# Markov Chains

The Flumphs from planet Seyst live in one of three states:
* Hungry
* Content
* Aggravated 

Based on the state of a Flumph on one day, you can predict the state on the day after. This could be represented as a 
table or matrix: <br>
### Initial chance-matrix and chance-vector
$H\leftarrow \mathbf{Hungry},\space C \leftarrow \mathbf{Content},\space A \leftarrow \mathbf{Aggravated}, \space
T \leftarrow \mathbf{Tomorrow},\space D \leftarrow \mathbf{Today}$ <br>

$\begin{array} {|c|c|c|c|} 
\hline D/T   & T_{H} & T_{C} & T_{A} \\
\hline D_{H} & 80\% &  10\% & 10\% \\
\hline D_{C} & 40\% & 50\% & 10\% \\
\hline D_{A} & 60\% & 20\% & 20\% \\ \hline
\end{array} $ <br> <br>

$M_{C} = \begin{bmatrix}0.80 & 0.40 &0.60 \\ 0.10 & 0.50 & 0.10\\ 0.10&0.10&0.20 \end{bmatrix}$ <br>

We also know that when a Flumph is born, it has a 10% chance to be hungry, a 70% chance to be content and a 20% chance 
to be aggravated. This can be turned into a vector. <br>
$\vec{c} = \begin{bmatrix} 0.10 \\ 0.70 \\ 0.20 \end{bmatrix}$ <br>

To calculate the odds for the day after a Flumph is born, one simply has to calculate the matrix-vector product between
these two. 
### The chances on the day after birth
$\vec{y} = \begin{bmatrix}0.80 & 0.40 &0.60 \\ 0.10 & 0.50 & 0.10\\ 0.10&0.10&0.20 \end{bmatrix} \cdot 
\begin{bmatrix} 0.10 \\ 0.70 \\ 0.20 \end{bmatrix}$
>$=\begin{bmatrix} 0.08+0.28+0.12\\
0.01+0.35+0.04\\
0.01+0.07+0.04\\ \end{bmatrix}$  

$\vec{y} = \begin{bmatrix} 0.48\\0.4\\0.12 \end{bmatrix}$ <br>
This means, that on the day after a Flumph is born, there is a 48% chance that the Flumph is hungry, a 40% chance that
the Flumph is content and a 12% chance that the Flumph is aggravated. <br>
Below, you will see this solved by the script:

In [10]:
def matrix_vector_product(activation, weights):
    output_layer = list()
    for w in weights:
        vector = 0.0
        for a in range(len(activation)):
            vector1 = float(float(activation[a]) * float(w[a]))
            vector += vector1
        output_layer.append(vector)
    return output_layer

chance_matrix_1 = [
    [0.8, 0.4, 0.6],
    [0.1, 0.5, 0.2],
    [0.1, 0.1, 0.2]]
chance_vector_1 = [0.10, 0.70, 0.20]

print("y =",matrix_vector_product(chance_vector_1, chance_matrix_1))

y = [0.48, 0.4, 0.12]


### Adding a 4th condition
A Flumph can also be upside-down! (https://en.wikipedia.org/wiki/Flumph). When a Flumph is upside-down on one day, the
chances of it being aggravated the next day are 30%, the chances of it being hungry is 30%, the chances of it being
content 10% and chances of it being upside-down are 30%. <br><br>
Due to adding a 4th condition, the chances in from the other conditions must also be altered. All of this results in the
following matrix:

$M_{C} = \begin{bmatrix} 
0.70 & 0.30 & 0.50 & 0.30 \\
0.10 & 0.60 & 0.10 & 0.30 \\
0.10 & 0.10 & 0.20 & 0.10 \\
0.10 & 0.10 & 0.20 & 0.30
 \end{bmatrix}$ <br><br>
Due to their floaty nature, it is likely that a Flumph is born upside-down. Because of this, the following chance-vector
is accurate at birth: <br>
$\vec{c} = \begin{bmatrix} 0.10\\0.50\\0.10\\0.30 \end{bmatrix}$ <br>
Assuming the same matrix-vector multiplication as before, the following chance-vector is accurate for the day after the 
Flumph is born, this time only using the function as defined above: 
### The chances on the day after birth

In [35]:
chance_matrix_2 = [
    [0.7, 0.3, 0.5, 0.3],
    [0.1, 0.5, 0.1, 0.3],
    [0.1, 0.1, 0.2, 0.1],
    [0.1, 0.1, 0.2, 0.3]]

chance_vector_2 = [0.1, 0.5, 0.1, 0.3]

print("y =",matrix_vector_product(chance_vector_2, chance_matrix_2))

y = [0.36, 0.36, 0.11000000000000001, 0.17]


$\vec{y} = \begin{bmatrix} 0.36\\0.41\\0.1\\0.17 \end{bmatrix}$

### Estimating the likeliest condition in the future
For both chance-matrices the nth chance-vector will be calculated through matrix-matrix multiplication.


In [36]:
import numpy as np
def matrix_matrix_product(matrix1, matrix2):
    matrix_product = list()
    for column in np.transpose(matrix2):
        matrix_product.append(matrix_vector_product(column, matrix1))
    matrix_product = np.array(matrix_product)
    return np.transpose(matrix_product)

In [9]:
n = 100

In [19]:
def get_nth(matrix, vector):
    matrix_c = matrix.copy()
    # n = global
    for i in range(n):
        matrix_c = matrix_matrix_product(matrix_c, matrix)
    y = matrix_vector_product(vector, matrix_c)
    # print(np.sum(y))
    # for i in y:
    #     print(round(i, 2), end="\\\\")
    return y

In [21]:
print("For chance-matrix 3-conditions {} days \n\ty =".format(n),get_nth(chance_matrix_1, chance_vector_1))
print()
print("For chance-matrix 4-conditions {} days \n\ty =".format(n),get_nth(chance_matrix_2, chance_vector_2))

For chance-matrix 3-conditions 100 days 
	y = [0.7037037037037046, 0.1851851851851853, 0.11111111111111127]

For chance-matrix 4-conditions 100 days 
	y = [0.537037037037037, 0.21296296296296308, 0.11111111111111113, 0.1388888888888889]


After 100 days, the chance-vectors stabilize to the following vectors:<br>
For 3 conditions: $\vec{y} = \begin{bmatrix}0.7\\0.19\\0.11 \end{bmatrix}$ <br><br>
For 4 conditions: $\vec{y} = \begin{bmatrix}0.54\\0.21\\0.11\\0.14 \end{bmatrix}$ <br>

From this, we can extrapolate that the likeliest condition for a Flumph by a large margin in both scenarios is hungry, 
followed up by content.

